numpy/core/src/multiarray/alloc.c

```diff
@@ -323,3 +324,300 @@ PyDataMem_RENEW(void *ptr, size_t size)
+        if (capsule == NULL) {
+            return NULL;
+        }
+    }
+    else {
+        capsule = PyCapsule_New(&default_handler, NULL, NULL);
+        if (capsule == NULL) {
+            return NULL;
+        }
+    }
+    token = PyContextVar_Set(current_handler, capsule);
+    Py_DECREF(capsule);
+    if (token == NULL) {
+        return NULL;
+    }
+    Py_DECREF(token);
+    return old_handler;
+#else
+    PyObject *p;
+    p = PyThreadState_GetDict();
+    if (p == NULL) {
+        return NULL;
+    }
+    old_capsule = PyDict_GetItemString(p, "current_allocator");
+    if (old_capsule == NULL) {
+        old_handler = &default_handler;
+    }
+    else {
+        old_handler = (PyDataMem_Handler *) PyCapsule_GetPointer(old_capsule, NULL);
+        Py_DECREF(old_capsule);
+        if (old_handler == NULL) {
+            return NULL;
+        }
+    }
+    if (handler != NULL) {
+        capsule = PyCapsule_New(handler, NULL, NULL);
+        if (capsule == NULL) {
+            return NULL;
+        }
+    }
+    else {
+        capsule = PyCapsule_New(&default_handler, NULL, NULL);
+        if (capsule == NULL) {
+            return NULL;
+        }
+    }
+    const int error = PyDict_SetItemString(p, "current_allocator", capsule);
+    Py_DECREF(capsule);
+    if (error) {
+        return NULL;
+    }
+    return old_handler;
+#endif
+}
+
+/*NUMPY_API
+ * Return the PyDataMem_Handler used by the PyArrayObject. If NULL, return
+ * the current global policy that will be used to allocate data
+ * for the next PyArrayObject. On failure, return NULL.
+ */
+NPY_NO_EXPORT const PyDataMem_Handler *
+PyDataMem_GetHandler(PyArrayObject *obj)
+{
+    PyObject *capsule;
+    PyDataMem_Handler *handler;
+#if (!defined(PYPY_VERSION_NUM) || PYPY_VERSION_NUM >= 0x07030600)
+    if (obj == NULL) {
+        if (PyContextVar_Get(current_handler, NULL, &capsule)) {
+            return NULL;
+        }
+        handler = (PyDataMem_Handler *) PyCapsule_GetPointer(capsule, NULL);
+        Py_DECREF(capsule);
+        return handler;
+    }
+#else
+    PyObject *p;
+    if (obj == NULL) {
+        p = PyThreadState_GetDict();
+        if (p == NULL) {
+            return NULL;
+        }
+        capsule = PyDict_GetItemString(p, "current_allocator");
+        if (capsule == NULL) {
+            handler = &default_handler;
+        }
+        else {
+            handler = (PyDataMem_Handler *) PyCapsule_GetPointer(capsule, NULL);
+            Py_DECREF(capsule);
+        }
+        return handler;
+    }
+#endif
+    /* Try to find a handler */
+    PyArrayObject *base = obj;
+    handler = PyArray_HANDLER(obj);
+    while (handler == NULL) {
+        base = (PyArrayObject*)PyArray_BASE(base);
+        /*
+         * If the base is an array which owns its own data, return its allocator.
+         */
+        if (base == NULL || ! PyArray_Check(base)) {
+            break;
+        }
+        handler = PyArray_HANDLER(base);
+    }
+    return handler;
```

Comment:
```
    /* If there's a handler, the array owns its own data */
    handler = PyArray_HANDLER(obj);
    if (handler == NULL) {
        /*
         * If the base is an array which owns its own data, return its allocator.
         */
        base = PyArray_BASE(obj);
        while (base != NULL && PyArray_Check(base)) {
            if (PyArray_CHKFLAGS((PyArrayObject *) base, NPY_ARRAY_OWNDATA)) {
                return PyArray_HANDLER(base);
            }
            base = PyArray_BASE(base);
        }
    }
    return handler;
```

Would something like this be more definite? I am not sure what are the possible cases, but this stops as soon as it finds the array that owns the data (and should have the right handler).

Reply:
On a second look, I think it should be just fine as is (only the comment is rather confusing, since there is no `NPY_ARRAY_OWNDATA` check).

Reply:
The current code traverses the bases and finds the first one with a handler. Yours traverses the bases, finds one that owns its data, and returns its handler unconditionally. I think they are equivalent, under the assumption that only arrays that own their data will have handlers. 

If we change that assumption, then I think the current code would be "more correct" since it will return the first handler it finds, which is probably what the user was looking for.

Reply:
True, the `If the base is an array which owns its own data, return its allocator.` comment could just be reworded (or deleted).

Reply:
On the other hand, if we finally decide to store the current/default allocator even if the array doesn't own its data (just in case the flag is changed afterwards, based on this [discussion](https://github.com/numpy/numpy/pull/17582/files#r694936774)) then the ownership check would be required.

Reply:
Hmm. I am not sure what we are looking for. If we have allocator policies `default`, `a`, and `b`, and somehow end up with this kind of a scenario:
```
hierarchy: narrary -> base -> base -> base-owning-data
policy:        --      b     default    a
```

I think the user would want `b`, since that was the last policy. I wonder if we should even be digging around in the `base` hierarchy, and shouldn't just return "whatever is in the context" if the current object's policy is `NULL`.

Reply:
```
    /*
     * If the obj is an array which owns its own data, return its allocator.
     */
    while (obj != NULL && PyArray_Check((PyObject *) obj)) {
        if (PyArray_CHKFLAGS(obj, NPY_ARRAY_OWNDATA)) {
            return PyArray_HANDLER(obj);
        }
        obj = (PyArrayObject *) PyArray_BASE(obj);
    }
    return NULL;
```

Probably this would be the easiest way to get it in that case.

Reply:
I think what matters most is to get the name of the one that actually allocated the memory addresses that we are accessing. Won't that be the data owner?

Reply:
If we only ever return the current object's allocator policy (and None if it is NULL), then the user could decide on their own how to traverse the `base` hierarchy. The `base` attribute and the `OWNDATA` flags are available from Python, they would not need C.

Reply:
You suggest keeping a simple `return PyArray_HANDLER(obj);` and changing `get_handler_name` to return `None` if `NULL`
```
    if (mem_handler == NULL) {
        Py_RETURN_NONE;
    }
````
right?

It sounds clear and doesn't try to guess the user's expected behavior. Tests will need modifications too. However, it would be problematic (for my project too) if it returns fake policies and not the ones used to allocate the underlying data (I am once again referencing that `NPY_ARRAY_OWNDATA` discussion).